### ISSUE TITLE
"Paren" is missing from ast

### ISSUE THREAD

**Issue by stuta** [mpeterv/luacheck#147]
I'm writing code generator using this great library. All I needed was to add ast and some other info to return of check.lua / check_or_throw().

```lua
return {
 events = events,
 per_line_options = per_line_options,
 line_lengths = utils.map(function(s) return #s end, chstate.source_lines),
 line_endings = line_endings
 ast = ast,
 comments = comments,
 code_lines = code_lines,
 source_lines = chstate.source_lines
}
```

Problem is if I parse ```a = (1+2)^3``` then there will not be any "Paren" -tags in ast. it is impossible to generate correct source from ast without this info.

**Comment by mpeterv** [comment]
Paren nodes are currently created only when they change meaning of the AST, so only around function calls and varargs at the end of expression list. If they were created for all real parens in source then the rest of checking code would have to unwrap them all the time. This can be reworked if it makes the parser easier to use from outside though. Are you trying  to parse some code with the luacheck parser, make some changes in AST, and then generate back code?

**Comment by stuta** [comment]
Yes, I'm changing ast and then generating fixed Lua code (for example adding needed require's and removing unused variables). I will also change Lua code to json that will be converted to javascript or other languages. I'm also generating automatic tests when module functions are added or parameters are changed.

With parens (1+2)^2 == 9, without parens 1+2^2 == 5 so parens are changing the meaning of code. Also boolean -statements need parens.

Maybe you could add option to include parens?

**Comment by mpeterv** [comment]
I think that if you are generating code for other languages then it may be a good idea to add parens automatically during generation instead. I'm not 100% sure about javascript but I wouldn't rely on other languages having exactly same operator precedence. You could simply wrap all operator expressions in parens, there are also ways to add parens only when required by target language precedence and associativity rules.

**Comment by stuta** [comment]
Language do not matter in this case. But it is impossible to add any parens when there is no information about parens in ast. Easy way is to generate 2 ast structuresin 2 runs, one without parens for easy check code generation and other with parents.

**Comment by mpeterv** [comment]
I forgot about this one, but since 0.23.0 parser does preserve parens, they are removed in a different stage now.
